File: packages/batch-submitter/src/da/span-batch-txs.ts
import { getBytes, toBeArray, toBigInt, zeroPadValue } from 'ethersv6'
import { Writer } from './types'
import { newSpanBatchTx } from './span-batch-tx'
import { encodeSpanBatchBits } from './utils'
import { L2Transaction, QueueOrigin } from '@localtest911/core-utils'

export class SpanBatchTxs {
  private totalBlockTxCount: number = 0
  private contractCreationBits: bigint = BigInt(0)
  private yParityBits: bigint = BigInt(0)
  private txSigs: SpanBatchSignature[] = []
  private txNonces: number[] = []
  private txGases: number[] = []
  private txTos: string[] = []
  private txDatas: Uint8Array[] = []
  private protectedBits: bigint = BigInt(0)
  private txTypes: number[] = []
  private totalLegacyTxCount: number = 0

  // metis extra fields
  private queueOriginBits: bigint = BigInt(0) // bitmap to save queue origins, 0 for sequencer, 1 for enqueue
  private l1TxOrigins: string[] = [] // l1 tx origins, only used for enqueue tx
  private txSeqSigs: SpanBatchSequencerSignature[] = []
  private seqYParityBits: bigint = BigInt(0)

  async addTxs(txs: L2Transaction[], chainId: bigint): Promise<void> {
    const offset = this.totalBlockTxCount
    for (let idx = 0; idx < txs.length; idx++) {
      const tx = txs[idx]

      // process legacy tx
      const txType = tx.type ?? 0
      if (!tx.type) {
        const protectedBit = tx.chainId ? BigInt(1) : BigInt(0)
        this.protectedBits |= protectedBit << BigInt(this.totalLegacyTxCount)
        this.totalLegacyTxCount++
      }

      if (tx.chainId && BigInt(tx.chainId) !== chainId) {
        throw new Error(
          `Protected tx has chain ID ${tx.chainId}, but expected chain ID ${chainId}`
        )
      }

      this.txSigs.push({
        r: tx?.signature.r ? BigInt(tx.signature.r) : BigInt(0),
        s: tx?.signature.s ? BigInt(tx.signature.s) : BigInt(0),
      })

      const contractCreationBit = tx.to ? BigInt(0) : BigInt(1)
      this.contractCreationBits |= contractCreationBit << BigInt(idx + offset)

      if (tx.to) {
        this.txTos.push(tx.to)
      }

      const yParityBit = BigInt(
        this.convertVToYParity(tx?.signature.v ?? 0, tx.type)
      )
      this.yParityBits |= yParityBit << BigInt(idx + offset)

      this.txNonces.push(Number(tx.nonce))
      this.txGases.push(Number(tx.gasLimit))
      this.txDatas.push(newSpanBatchTx(tx).marshalBinary())
      this.txTypes.push(txType)

      // append metis extra fields
      this.queueOriginBits |=
        BigInt(tx.queueOrigin === QueueOrigin.Sequencer) << BigInt(idx + offset)
      if (tx.queueOrigin !== QueueOrigin.Sequencer) {
        this.l1TxOrigins.push(tx.l1TxOrigin)
      }
      this.txSeqSigs.push({
        r: tx.seqR ? toBigInt(tx.seqR) : BigInt(0),
        s: tx.seqS ? toBigInt(tx.seqS) : BigInt(0),
      })
      const seqYParityBit = tx.seqV ? toBigInt(tx.seqV) : BigInt(0)
      this.seqYParityBits |= seqYParityBit << BigInt(idx + offset)
    }

    this.totalBlockTxCount += txs.length
  }

  encode(): Uint8Array {
    const writer = new Writer()
    this.encodeContractCreationBits(writer)
    this.encodeYParityBits(writer)
    this.encodeTxSigsRS(writer)
    this.encodeTxTos(writer)
    this.encodeTxDatas(writer)
    this.encodeTxNonces(writer)
    this.encodeTxGases(writer)
    this.encodeProtectedBits(writer)

    // encode metis extra fields
    this.encodeQueueOriginBits(writer)
    this.encodeSeqYParityBits(writer)
    this.encodeTxSeqSigsRS(writer)
    this.encodeL1TxOrigins(writer)
    return writer.getData()
  }

  private convertVToYParity(v: number, txType: number): number {
    switch (txType) {
      case 0: // Legacy
        return this.isProtectedV(v, txType)
          ? Number((v - 35) & 1)
          : Number(v - 27)
      case 1: // AccessList
      case 2: // DynamicFee
        return Number(v)
      default:
        throw new Error(`Invalid tx type: ${txType}`)
    }
  }

  private isProtectedV(v: number, txType: number): boolean {
    return txType === 0 ? v !== 27 && v !== 28 : true
  }

  private encodeQueueOriginBits(writer: Writer): void {
    encodeSpanBatchBits(writer, this.totalBlockTxCount, this.queueOriginBits)
  }

  private encodeContractCreationBits(writer: Writer): void {
    encodeSpanBatchBits(
      writer,
      this.totalBlockTxCount,
      this.contractCreationBits
    )
  }

  private encodeYParityBits(writer: Writer): void {
    encodeSpanBatchBits(writer, this.totalBlockTxCount, this.yParityBits)
  }

  private encodeTxSigsRS(writer: Writer): void {
    for (const txSig of this.txSigs) {
      writer.writeBytes(zeroPadValue(toBeArray(txSig.r), 32))
      writer.writeBytes(zeroPadValue(toBeArray(txSig.s), 32))
    }
  }

  private encodeTxTos(writer: Writer): void {
    for (const txTo of this.txTos) {
      writer.writeBytes(getBytes(txTo))
    }
  }

  private encodeL1TxOrigins(writer: Writer): void {
    for (const l1TxOrigin of this.l1TxOrigins) {
      writer.writeBytes(getBytes(l1TxOrigin))
    }
  }

  private encodeTxDatas(writer: Writer): void {
    for (const txData of this.txDatas) {
      writer.writeBytes(txData)
    }
  }

  private encodeTxNonces(writer: Writer): void {
    for (const txNonce of this.txNonces) {
      writer.writeVarInt(txNonce)
    }
  }

  private encodeTxGases(writer: Writer): void {
    for (const txGas of this.txGases) {
      writer.writeVarInt(txGas)
    }
  }

  private encodeProtectedBits(writer: Writer): void {
    encodeSpanBatchBits(writer, this.totalLegacyTxCount, this.protectedBits)
  }

  private encodeTxSeqSigsRS(writer: Writer): void {
    for (const txSeqSig of this.txSeqSigs) {
      writer.writeBytes(zeroPadValue(toBeArray(txSeqSig.r), 32))
      writer.writeBytes(zeroPadValue(toBeArray(txSeqSig.s), 32))
    }
  }

  private encodeSeqYParityBits(writer: Writer): void {
    encodeSpanBatchBits(writer, this.totalBlockTxCount, this.seqYParityBits)
  }
}

interface SpanBatchSignature {
  r: bigint
  s: bigint
}

interface SpanBatchSequencerSignature {
  r: bigint
  s: bigint
}

File: go/op-program/client/derive/span_batch_txs.go
package derive

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"math/big"

	"github.com/MetisProtocol/mvm/l2geth/common"
	"github.com/MetisProtocol/mvm/l2geth/common/hexutil"
	"github.com/MetisProtocol/mvm/l2geth/core/types"
	"github.com/MetisProtocol/mvm/l2geth/rlp"
	dtl "github.com/MetisProtocol/mvm/l2geth/rollup"

	"github.com/ethereum-optimism/optimism/op-node/rollup/derive"
	"github.com/holiman/uint256"
)

type spanBatchTxs struct {
	// this field must be manually set
	totalBlockTxCount uint64

	// 8 fields
	contractCreationBits *big.Int // standard span-batch bitlist
	yParityBits          *big.Int // standard span-batch bitlist
	txSigs               []spanBatchSignature
	txNonces             []uint64
	txGases              []uint64
	txTos                []common.Address
	txDatas              []hexutil.Bytes
	protectedBits        *big.Int // standard span-batch bitlist

	// metis fields
	queueOriginBits *big.Int
	l1TxOrigins     []common.Address
	txSeqSigs       []spanBatchSignature
	seqYParityBits  *big.Int

	l1BlockNumbers []uint64
	l1Timestamps   []uint64
	blockTxCounts  []uint64

	totalEnqueueTxCount uint64

	// intermediate variables which can be recovered
	txTypes            []int
	totalLegacyTxCount uint64
}

type spanBatchSignature struct {
	v uint64
	r *uint256.Int
	s *uint256.Int
}

func (btx *spanBatchTxs) decodeContractCreationBits(r *bytes.Reader) error {
	if btx.totalBlockTxCount > derive.MaxSpanBatchElementCount {
		return derive.ErrTooBigSpanBatchSize
	}
	bits, err := decodeSpanBatchBits(r, btx.totalBlockTxCount)
	if err != nil {
		return fmt.Errorf("failed to decode contract creation bits: %w", err)
	}
	btx.contractCreationBits = bits
	return nil
}

func (btx *spanBatchTxs) decodeProtectedBits(r *bytes.Reader) error {
	if btx.totalLegacyTxCount > derive.MaxSpanBatchElementCount {
		return derive.ErrTooBigSpanBatchSize
	}
	bits, err := decodeSpanBatchBits(r, btx.totalLegacyTxCount)
	if err != nil {
		return fmt.Errorf("failed to decode protected bits: %w", err)
	}
	btx.protectedBits = bits
	return nil
}

func (btx *spanBatchTxs) decodeQueueOriginBits(r *bytes.Reader) error {
	if btx.totalLegacyTxCount > derive.MaxSpanBatchElementCount {
		return derive.ErrTooBigSpanBatchSize
	}
	bits, err := decodeSpanBatchBits(r, btx.totalBlockTxCount)
	if err != nil {
		return fmt.Errorf("failed to decode queue origin bits: %w", err)
	}
	btx.queueOriginBits = bits
	return nil
}

func (btx *spanBatchTxs) decodeSeqYParityBits(r *bytes.Reader) error {
	if btx.totalLegacyTxCount > derive.MaxSpanBatchElementCount {
		return derive.ErrTooBigSpanBatchSize
	}
	bits, err := decodeSpanBatchBits(r, btx.totalBlockTxCount)
	if err != nil {
		return fmt.Errorf("failed to decode seq y parity bits: %w", err)
	}
	btx.seqYParityBits = bits
	return nil
}

func (btx *spanBatchTxs) contractCreationCount() (uint64, error) {
	if btx.contractCreationBits == nil {
		return 0, errors.New("dev error: contract creation bits not set")
	}
	var result uint64 = 0
	for i := 0; i < int(btx.totalBlockTxCount); i++ {
		bit := btx.contractCreationBits.Bit(i)
		if bit == 1 {
			result++
		}
	}
	return result, nil
}

func (btx *spanBatchTxs) enqueueTxCount() (uint64, error) {
	if btx.queueOriginBits == nil {
		return 0, errors.New("dev error: contract creation bits not set")
	}
	var result uint64 = 0
	for i := 0; i < int(btx.totalBlockTxCount); i++ {
		bit := btx.queueOriginBits.Bit(i)
		if bit == 1 {
			result++
		}
	}
	return result, nil
}

func (btx *spanBatchTxs) decodeYParityBits(r *bytes.Reader) error {
	bits, err := decodeSpanBatchBits(r, btx.totalBlockTxCount)
	if err != nil {
		return fmt.Errorf("failed to decode y-parity bits: %w", err)
	}
	btx.yParityBits = bits
	return nil
}

func (btx *spanBatchTxs) decodeTxSigsRS(r *bytes.Reader) error {
	var txSigs []spanBatchSignature
	var sigBuffer [32]byte
	for i := 0; i < int(btx.totalBlockTxCount); i++ {
		var txSig spanBatchSignature
		_, err := io.ReadFull(r, sigBuffer[:])
		if err != nil {
			return fmt.Errorf("failed to read tx sig r: %w", err)
		}
		txSig.r, _ = uint256.FromBig(new(big.Int).SetBytes(sigBuffer[:]))
		_, err = io.ReadFull(r, sigBuffer[:])
		if err != nil {
			return fmt.Errorf("failed to read tx sig s: %w", err)
		}
		txSig.s, _ = uint256.FromBig(new(big.Int).SetBytes(sigBuffer[:]))
		txSigs = append(txSigs, txSig)
	}
	btx.txSigs = txSigs
	return nil
}

func (btx *spanBatchTxs) decodeTxSeqSigsRS(r *bytes.Reader) error {
	var txSeqSigs []spanBatchSignature
	var sigBuffer [32]byte
	for i := 0; i < int(btx.totalBlockTxCount); i++ {
		var txSig spanBatchSignature
		_, err := io.ReadFull(r, sigBuffer[:])
		if err != nil {
			return fmt.Errorf("failed to read tx sig r: %w", err)
		}
		txSig.r, _ = uint256.FromBig(new(big.Int).SetBytes(sigBuffer[:]))
		_, err = io.ReadFull(r, sigBuffer[:])
		if err != nil {
			return fmt.Errorf("failed to read tx sig s: %w", err)
		}
		txSig.s, _ = uint256.FromBig(new(big.Int).SetBytes(sigBuffer[:]))
		txSeqSigs = append(txSeqSigs, txSig)
	}
	btx.txSeqSigs = txSeqSigs
	return nil
}

func (btx *spanBatchTxs) decodeTxNonces(r *bytes.Reader) error {
	var txNonces []uint64
	for i := 0; i < int(btx.totalBlockTxCount); i++ {
		txNonce, err := binary.ReadUvarint(r)
		if err != nil {
			return fmt.Errorf("failed to read tx nonce: %w", err)
		}
		txNonces = append(txNonces, txNonce)
	}
	btx.txNonces = txNonces
	return nil
}

func (btx *spanBatchTxs) decodeTxGases(r *bytes.Reader) error {
	var txGases []uint64
	for i := 0; i < int(btx.totalBlockTxCount); i++ {
		txGas, err := binary.ReadUvarint(r)
		if err != nil {
			return fmt.Errorf("failed to read tx gas: %w", err)
		}
		txGases = append(txGases, txGas)
	}
	btx.txGases = txGases
	return nil
}

func (btx *spanBatchTxs) decodeTxTos(r *bytes.Reader) error {
	var txTos []common.Address
	txToBuffer := make([]byte, common.AddressLength)
	contractCreationCount, err := btx.contractCreationCount()
	if err != nil {
		return err
	}
	for i := 0; i < int(btx.totalBlockTxCount-contractCreationCount); i++ {
		_, err := io.ReadFull(r, txToBuffer)
		if err != nil {
			return fmt.Errorf("failed to read tx to address: %w", err)
		}
		txTos = append(txTos, common.BytesToAddress(txToBuffer))
	}
	btx.txTos = txTos
	return nil
}

func (btx *spanBatchTxs) decodeL1TxOrigins(r *bytes.Reader) error {
	var txOrigins []common.Address
	l1TxOriginsBuffer := make([]byte, common.AddressLength)
	enqueueTxCount, err := btx.enqueueTxCount()
	if err != nil {
		return err
	}
	for i := 0; i < int(enqueueTxCount); i++ {
		_, err := io.ReadFull(r, l1TxOriginsBuffer)
		if err != nil {
			return fmt.Errorf("failed to read tx to address: %w", err)
		}
		txOrigins = append(txOrigins, common.BytesToAddress(l1TxOriginsBuffer))
	}
	btx.l1TxOrigins = txOrigins
	return nil
}

func (btx *spanBatchTxs) decodeTxDatas(r *bytes.Reader) error {
	var txDatas []hexutil.Bytes
	var txTypes []int
	// Do not need txDataHeader because RLP byte stream already includes length info
	for i := 0; i < int(btx.totalBlockTxCount); i++ {
		txData, txType, err := derive.ReadTxData(r)
		if err != nil {
			return err
		}
		txDatas = append(txDatas, txData)
		txTypes = append(txTypes, txType)
		if txType == 0 {
			btx.totalLegacyTxCount++
		}
	}
	btx.txDatas = txDatas
	btx.txTypes = txTypes
	return nil
}

func (btx *spanBatchTxs) recoverV(chainID *big.Int) error {
	if len(btx.txTypes) != len(btx.txSigs) {
		return errors.New("tx type length and tx sigs length mismatch")
	}
	if btx.protectedBits == nil {
		return errors.New("dev error: protected bits not set")
	}
	protectedBitsIdx := 0
	for idx, txType := range btx.txTypes {
		var v uint64
		queueOrigin := btx.queueOriginBits.Bit(idx)
		if types.QueueOrigin(queueOrigin) == types.QueueOriginL1ToL2 {
			// for enqueue tx we do not need to parse v
			v = 0
		} else {
			bit := uint64(btx.yParityBits.Bit(idx))
			switch txType {
			case 0:
				protectedBit := btx.protectedBits.Bit(protectedBitsIdx)
				protectedBitsIdx++
				if protectedBit == 0 {
					v = 27 + bit
				} else {
					// EIP-155
					v = chainID.Uint64()*2 + 35 + bit
				}
			default:
				return fmt.Errorf("invalid tx type: %d", txType)
			}
		}

		btx.txSigs[idx].v = v
	}
	return nil
}

func (btx *spanBatchTxs) decode(r *bytes.Reader) error {
	if err := btx.decodeContractCreationBits(r); err != nil {
		return err
	}
	if err := btx.decodeYParityBits(r); err != nil {
		return err
	}
	if err := btx.decodeTxSigsRS(r); err != nil {
		return err
	}
	if err := btx.decodeTxTos(r); err != nil {
		return err
	}
	if err := btx.decodeTxDatas(r); err != nil {
		return err
	}
	if err := btx.decodeTxNonces(r); err != nil {
		return err
	}
	if err := btx.decodeTxGases(r); err != nil {
		return err
	}
	if err := btx.decodeProtectedBits(r); err != nil {
		return err
	}
	if err := btx.decodeQueueOriginBits(r); err != nil {
		return err
	}
	if err := btx.decodeSeqYParityBits(r); err != nil {
		return err
	}
	if err := btx.decodeTxSeqSigsRS(r); err != nil {
		return err
	}
	if err := btx.decodeL1TxOrigins(r); err != nil {
		return err
	}
	return nil
}

func (btx *spanBatchTxs) fullTxs(chainID *big.Int, startBlock uint64, enqueue map[uint64]*dtl.Enqueue) (types.Transactions, error) {
	var txs types.Transactions
	toIdx := 0
	enqueueTxIdx := 0
	blockIndex := uint64(0)
	for idx := 0; idx < int(btx.totalBlockTxCount); idx++ {
		idxBlockUpperBound := uint64(0)
		for i := uint64(0); i <= blockIndex; i++ {
			idxBlockUpperBound += btx.blockTxCounts[i]
		}

		if uint64(idx) >= idxBlockUpperBound {
			blockIndex++
		}

		var stx spanBatchTx
		if err := stx.UnmarshalBinary(btx.txDatas[idx]); err != nil {
			return nil, err
		}
		nonce := btx.txNonces[idx]
		gas := btx.txGases[idx]
		var to *common.Address = nil
		bit := btx.contractCreationBits.Bit(idx)
		if bit == 0 {
			if len(btx.txTos) <= toIdx {
				return nil, errors.New("tx to not enough")
			}
			to = &btx.txTos[toIdx]
			toIdx++
		}

		var (
			l1TxOrigin  common.Address
			enqueueData []byte
		)
		queueIndex := uint64(0)
		queueOrigin := types.QueueOrigin(btx.queueOriginBits.Bit(idx))
		isEnqueueTx := queueOrigin == types.QueueOriginL1ToL2
		if isEnqueueTx {
			if len(btx.l1TxOrigins) <= enqueueTxIdx {
				return nil, errors.New("l1 tx origins not enough")
			}
			l1TxOrigin = btx.l1TxOrigins[enqueueTxIdx]
			queueIndex = nonce
			payload := enqueue[queueIndex]
			if payload != nil {
				enqueueData = *payload.Data
			}
			enqueueTxIdx++
		}

		r := btx.txSigs[idx].r
		s := btx.txSigs[idx].s

		tx, err := stx.convertToFullTx(nonce, gas, to, chainID, r, s, byte(getBit(btx.yParityBits, idx)),
			getBit(btx.protectedBits, idx) == 1, isEnqueueTx, enqueueData)
		if err != nil {
			return nil, err
		}

		// Note: set l2 tx to encode only the data part
		var rawTx []byte
		if isEnqueueTx {
			rawTx = enqueueData
		} else {
			tx.SetL2Tx(2)
			rawTx, err = rlp.EncodeToBytes(tx)
			if err != nil {
				return nil, err
			}
			tx.SetL2Tx(0)
		}

		batchBlockIndex := startBlock + blockIndex - 1

		txMeta := &types.TransactionMeta{
			L1BlockNumber:   new(big.Int).SetUint64(btx.l1BlockNumbers[blockIndex]),
			L1Timestamp:     btx.l1Timestamps[blockIndex],
			L1MessageSender: &l1TxOrigin,
			QueueOrigin:     queueOrigin,
			Index:           &batchBlockIndex,
			QueueIndex:      &queueIndex,
			RawTransaction:  rawTx,
			R:               btx.txSeqSigs[idx].r.ToBig(),
			S:               btx.txSeqSigs[idx].s.ToBig(),
			V:               big.NewInt(int64(btx.seqYParityBits.Bit(idx))),
		}
		tx.SetTransactionMeta(txMeta)
		txs = append(txs, tx)
	}
	return txs, nil
}

func getBit(bits *big.Int, index int) uint {
	return bits.Bit(index)
}

func isProtectedV(v uint64, txType int) bool {
	if txType == 0 {
		// if EIP-155 applied, v = 2 * chainID + 35 + yParity
		return v != 27 && v != 28
	}
	// every non legacy tx are protected
	return true
}
